### ISSUE TITLE
Crash on UIKeyboardNotification on glView on iOS platform

### ISSUE THREAD

**Issue by AlwynC** [cocos2d/cocos2d-x#15733]
When cocos 2d-x is integrated on top of UIKit,that is,we have a few native iOS view controllers and then move into cocos 2d-x screen,the application works fine.Now,when we move back to iOS screen from the cocos 2d-x screen by invoking Director::getInstance()->end(),without exiting the application,we are able to transition back to iOS screen from cocos 2d-x screen without problems.Supposing there is a text field in the iOS screen and we tap on the text field after transitioning from the cocos 2d-x screen,the application crashes when the keyboard is to be shown up for the textfield in the iOS screen.This is because the UIKeyboardNotification which was set on the glView of the cocos 2d-x screen has not been removed when releasing the glView during Director::getInstance()->end() and the UIKeyboardNotification is still active and called on the released glView.On inspecting the code in CCEAGLView-ios.mm file,in the dealloc method,the UIKeyboardNotification is being removed and other cleanup of the glview is done.But when Director::getInstance()->end() is invoked it does not invoke the dealloc method of CCEAGLView at all.Hence the UIKeyboardNotification is still active even though the glView is released.Could the Director::getInstance()->end() functionality be improvised so that the dealloc method of CCEAGLView-ios.mm is invoked implicitly?

**Comment by minggo** [comment]
@AlwynC Could you please provide a test case? Thanks.

**Comment by AlwynC** [comment]
@minggo Is it fine if I share a sample application which reproduces the crash?

**Comment by minggo** [comment]
@AlwynC Yep, thanks.

**Comment by AlwynC** [comment]
@minggo The sample application using cocos2d-x v3.11.1 reproducing the crash is ready.How do I share it?

**Comment by minggo** [comment]
@AlwynC it is better if you can for cocos2d-x and create a branch for it, or you can paste some codes that i can used it in `cpp-empty-test` to reproduce it

**Comment by AlwynC** [comment]
@minggo I've pushed the sample application reproducing the crash on my forked cocos2d-x Github repository https://github.com/AlwynC/cocos2d-x/tree/AlwynC_iOSKeyboardCrash on the Alwyn_iOSKeyboardCrash branch.The folder for the application is added as MyGame within the repository folder.Please have a look.

**Comment by AlwynC** [comment]
The crash occurs due to creating a CCEAGLView property with strong semantics in RootViewController.mm as @property(strong,nonatomic) CCEAGLView *eaglView.Since manual reference counting is followed,if we do not manually release the CCEAGLView property using self.eaglView = nil in RootViewController.mm file,the dealloc method of CCEAGLView-ios.mm is not called since the CCEAGLView instance in RootViewController.mm is still not released and the UIKeyboardNotification on the CCEAGLView is also active.Hence when we navigate back to iOS screen from the cocos2d-x scene by invoking Director::getInstance()->end() and tap on the textfield in the iOS screen,there is a crash when keyboard is to be shown up for editing the text field in iOS screen.

**Comment by AlwynC** [comment]
To fix this,we can perform the set up for CCEAGLView in RootViewController.mm in the viewDidLoad method,in the same way as I have done in the sample project and manually release the CCEAGLView in viewDidDisappear method.In this way,there will be no crash on the iOS screen's textfield.I can provide the fix in the next commit if you say so.

**Comment by minggo** [comment]
@AlwynC Thanks. I will think about it when i have time. I am busy right now.

**Comment by AlwynC** [comment]
@minggo Sure.I've anyways added the fix for the crash on the sample application on https://github.com/AlwynC/cocos2d-x/tree/AlwynC_iOSKeyboardCrash_Fix and refined the code on the actual cocos2d-x cpp templates for iOS on https://github.com/AlwynC/cocos2d-x/tree/refine_cpp_template_for_iOS. Please do have a look and do get back to me if you have any concerns.Thanks.

**Comment by minggo** [comment]
> The crash occurs due to creating a CCEAGLView property with strong semantics in RootViewController.mm as @property(strong,nonatomic) CCEAGLView *eaglView.Since manual reference counting is followed

cocos2d-x doesn't create a CCEAGLView property in RootViewController.mm. So it is your code created it? I think you should be take care of it for your code.

**Comment by AlwynC** [comment]
@minggo Yes cocos2d-x does not create a CCEAGLView property in RootViewController.mm.It is created as local variable in AppController.mm and the RootViewController is presented as the first scene of the application by default in cocos2d-x.That is how it is in the cpp-templates for iOS.But generally,when we build a game using cocos2d-x on iOS,we normally have a few native iOS screens on top of which we present the cocos2d-x scene.In that case,we will load the iOS screen in AppController.mm instead of RootViewController and we have to create a CCEAGLView as a property on RootViewController.mm with strong semantics so that we can use the property elsewhere.If we fail to take care of releasing the CCEAGLView property on RootViewController.mm when quitting the game scene and moving back to iOS screen,the application will crash on textfield in iOS screen and it is very difficult to find out the cause of the crash.Hence we can refine the code in the cocos2d-x cpp-templates folder to set up the CCEAGLView as a property and releasing it in RootViewController.mm rather than AppController.mm as I have done in  https://github.com/AlwynC/cocos2d-x/tree/refine_cpp_template_for_iOS so as to prevent any crash in future.Please do have look.

**Comment by minggo** [comment]
@AlwynC May be we should uncomment [these codes](https://github.com/cocos2d/cocos2d-x/blob/v3/cocos/platform/ios/CCGLViewImpl-ios.mm#L118) to release eaglview. But i don't know why it is commented. And i think it is something related with using cocos2d-x with cocoa.

**Comment by AlwynC** [comment]
@minggo I tried out the solution you suggested above to release eaglview for two cases-
1.When CCEAGLView is created as a local variable - the above solution leads to a crash since a release is being sent to an already autoreleased instance of CCEAGLView.

2.When CCEAGLView is created as a property with strong/retain semantics as @property(strong,nonatomic) CCEAGLview *eaglView or @property(retain,nonatomic) CCEAGLView *eaglView - the above solution works and there is no crash.

Since there was a crash in the first case with uncommenting the codes you mentioned above,we cannot uncomment the codes you mentioned above.To prevent a crash for both the above cases,I refined the code on CCEAGLView-ios.mm and removed extra release in CCGLViewImpl-ios.mm.I've pushed the code on https://github.com/AlwynC/cocos2d-x on the AlwynC_iOSKeyboardCrashFix branch.With this approach releasing the CCEAGLView instance is more effective and works very well for both above cases 1 and  2 with no crashes.Please do have a look.You may let me know your views on this.Thanks.

**Comment by Shulepov** [comment]
I think issue in current implementation is next:
`_viewController = [[RootViewController alloc] initWithNibName:nil bundle:nil];`

There is no `release` in dealloc or `autorelease` of `_viewController`. So it's view also not released. After fixing that and trying to call `Director::end` - CCEAGLView was successfully deallocated.

Issue with releasing `eaglView` property belongs only to your code... And i think it's obvious that it should be released. Actually instead of strongly assigning it you can implement readonly property like:

```
- (CCEAGLView *)eaglView {
     //return (CCEAGLView *) cocos2d::Director::getInstance()->getOpenGLView()->getEAGLView();
     return (CCEAGLView *)self.view;
}
```

Also there is one note:
when user touches a screen there are created `UITouch` objects, which keep strongly `UIView` that them touch. And those touches are not destroyed until next touch. So actually CCEAGLView will be released after next touch.

**Comment by Shulepov** [comment]
@minggo this `release` line is commented properly, because `GLViewImpl` not retains it, so it shouldn't release it.

**Comment by AlwynC** [comment]
@Shulepov Yes issue with releasing eaglView strong/retain property belongs to my code as you said.But if I somehow miss out on releasing it when closing the game scene,then the cocos2d-x engine should be able to handle it. You've mentioned something about touches.This issue is not related with touches and releasing CCEAGLView after next touch.It's only about releasing a strong,retained CCEAGLView in RootViewController.mm.While creating CCEAGLView instance it's autoreleased.But if I have a strong/retain proeprty of CCEAGLview in my viewcontoller ,it is not autoreleased but I have to handle the release in my code when dismissing my viewcontroller and moving back to iOS screen.So if I've missed out on releasing the CCEAGLView which I've retained somehow,the cocos2d-x engine has to take care of this,that's what I've suggested in my code above.

**Comment by Shulepov** [comment]
Actually, i think it may be a good idea to move CCEAGLView initialization into RootViewController.

But want to make few comments:
- is there a real need to create private property for `eaglView`? Even in that case it can be released in `dealloc` as usual. If you'll follow that rule - things that was retained should be released - then there are wouldn't appear any issues. Or you can enable ARC support :)
- currently `CCEAGLView` added as a root view, so it may be better to do the work inside `loadView` and assign it to `self.view`

**Comment by AlwynC** [comment]
@Shulepov That's what I'd suggested before-Moving CCEAGLView initialization to the root view controller from the AppController.There are many ways to initialize the CCEAGLView either as a private property,or by local initialization etc. which is decided by the developer who is building his application using cocos2d-x.What my point is that for whichever approach the developer follows to create the CCEAGLView,and if for whatever reason there are memory issues occurring as a result of that,cocos2d-x should be able to handle it internally,the suggestions for which I've provided above.And regarding ARC support,it may not be possible since there are lot of void\* pointer casts for converting between CCEAGLView to GLViewImpl which will create problems when ARC is enabled.

**Comment by Shulepov** [comment]
@AlwynC i don't think that cocos2d-x can properly handle memory issues introduced by the developers. To find them should be used profiling instruments. Though cocos2d-x can try to hide some issues that can appear by making cleanup of CCEAGLView when `Director::end` called - so memory leaks still will be there, but without crashes. But that's also questionable.

Btw moving to ARC is quite easy - https://github.com/cocos2d/cocos2d-x/pull/13937

I like idea moving CCEAGLView initialization to RootViewController, but it has nothing with memory issues.

**Comment by AlwynC** [comment]
@Shulepov I'll try out the ARC transition for iOS as mentioned in #13937 ,and see if that would be useful to handle memory issues introduced by developers.

**Comment by minggo** [comment]
@AlwynC Yep, cocos2d-x can not resolve memory issues caused by developer codes.

**Comment by AlwynC** [comment]
@minggo @Shulepov I tried with ARC enabled as mentioned in #13937 but still there is a crash when CCEAGLView is created as a strong property and not released in RootViewController.mm.Hence if strong/retain properties are created by the developer in RootViewController,the developer has to compulsorily perform the cleanup for the strongly retained property by himself when closing the cocos2d-x game scene since cocos2d-x cannot resolve such memory issues.

**Comment by AlwynC** [comment]
@minggo I have a suggestion.Can we have a more detailed assertion during the crash due to such memory issues which will notify the developer that he has not released something which he has retained as part of his code?I think this way debugging such memory issues will be easier for the developer.

**Comment by minggo** [comment]
@AlwynC I have not idea how to add the asset codes.

**Comment by AlwynC** [comment]
@minggo It's just CCAssert.I've added the assertion for this crash on https://github.com/AlwynC/cocos2d-x/tree/AlwynC_iOSKeyboardCrashFix. Could you please have look?Also,I think we can move the CCEAGLView setup from AppController.mm to RootViewController.mm as I've done in https://github.com/AlwynC/cocos2d-x/tree/refine_cpp_template_for_iOS. Please have a look.

**Comment by minggo** [comment]
> https://github.com/AlwynC/cocos2d-x/tree/AlwynC_iOSKeyboardCrashFix.Could you please have look?

@AlwynC Sorry, i can not open the link. Could you just show me the commit?

> Also,I think we can move the CCEAGLView setup from AppController.mm to RootViewController.mm as I've done in https://github.com/AlwynC/cocos2d-x/tree/refine_cpp_template_for_iOS. Please have a look.

Sounds good to me, will you send a PR for it?

**Comment by AlwynC** [comment]
@minggo Here is the commit link. https://github.com/AlwynC/cocos2d-x/commit/b26ac5ee47bb5ac5b3126d7a967168675398c930.

**Comment by Shulepov** [comment]
I think main issue here was of using `CCEAGLViewImpl` from `CCEAGLView` after `Director::end` was called. And usually it's accessed via `Director::getInstance()->getOpenGLView()`. Possibly some assertions may be added to `getOpenGLView` to check that it's not null.

P.S.
@AlwynC sorry for stupid question, but have you enabled ARC in build settings after applying those changes?

**Comment by AlwynC** [comment]
@minggo Submitted a pull request for cpp-template refining at #15792 .

**Comment by AlwynC** [comment]
@Shulepov Turns out that I had not enabled ARC in build settings after applying the changes in #13937 earlier.Hence I still got the crash.I tested it out now with ARC enabled in build settings along with the ARC changes in code as you've mentioned.And I'm very happy to say it works pretty good!!!Even if I don't release the strongly retained CCEAGLView in RootViewController.mm manually,with ARC enabled it works without any crashes.Thank you for #13937 .Also,as you said maybe adding assertions on accessing openGLView of Director may be more useful.

**Comment by AlwynC** [comment]
@minggo @Shulepov This issue has been fixed by pull request #15792 (for non ARC) and pull request #13937(with ARC support).Hence closing this issue.
